Editor/EditorCore/DrawShapeTool.cs

```diff
@@ -0,0 +1,380 @@
+        /// </summary>
+        /// <param name="diff">Difference between point A and point B</param>
+        /// <returns></returns>
+        Quaternion ToRotationAngles(Vector3 diff)
+        {
+            Vector3 angles = Vector3.zero;
+            if (diff.y < 0)
+            {
+                angles.z = -180f;
+            }
+            if (diff.z < 0)
+            {
+                angles.y = -180f;
+            }
+            return Quaternion.Euler(angles);
+        }
+
+        void SetWidthAndDepth(Event evt)
+        {
+            switch (evt.type)
+            {
+                case EventType.MouseDrag:
+                    {
+                        m_IsDragging = true;
+                        Ray ray = HandleUtility.GUIPointToWorldRay(evt.mousePosition);
+                        float distance;
+
+                        if (m_Plane.Raycast(ray, out distance))
+                        {
+                            m_OppositeCorner = ray.GetPoint(distance);
+                            m_HeightCorner = m_OppositeCorner;
+                            var diff = m_OppositeCorner - m_Origin;
+                            RebuildShape();
+                            SceneView.RepaintAll();
+                        }
+                        break;
+                    }
+
+                case EventType.MouseUp:
+                    {
+                        if (!m_IsDragging)
+                        {
+                            m_IsDragging = true;
+                            Ray ray = HandleUtility.GUIPointToWorldRay(evt.mousePosition);
+                            float distance;
+
+                            if (m_Plane.Raycast(ray, out distance))
+                            {
+                                var pos = ray.GetPoint(distance);
+                                CancelShape();
+                                var shape = CreateLastShape(Vector3.one * distance / 5f);
+                                shape.transform.position = pos;
+                            }
+                        }
+                        else if (Vector3.Distance(m_OppositeCorner, m_Origin) < .1f)
+                            CancelShape();
+                        else
+                            AdvanceInputState();
+                        break;
+
+                    }
+            }
+        }
+
+        public static ProBuilderMesh CreateLastShape(Vector3 defaultSize)
+        {
+            if (s_Size == Vector3.zero)
+                s_Size = defaultSize;
+            var type = activeShapeType;
+            var shape = ShapeGenerator.CreateShape(type).GetComponent<ShapeComponent>();
+            ShapeParameters.SetToLastParams(ref shape.shape);
+            UndoUtility.RegisterCreatedObjectUndo(shape.gameObject, "Create Shape");
+
+            Bounds bounds = new Bounds(Vector3.zero, s_Size);
+            shape.Rebuild(bounds, Quaternion.identity);
+            shape.mesh.SetPivot(PivotLocation.Center);
```

Comment:
This should be handled in `EditorUtility.InitObject`. The `PivotLocation` is a user preference, and not always center.